gwcs/wcs.py

```diff
@@ -634,3 +641,273 @@ def fix_inputs(self, fixed):
+        pdegree = fit_poly_x.degree
+        if pdegree > 1:
+            hdr['a_order'] = pdegree
+            hdr['b_order'] = pdegree
+            _store_2D_coefficients(hdr, sip_poly_x, 'A')
+            _store_2D_coefficients(hdr, sip_poly_y, 'B')
+            if max_inv_pix_error:
+                _store_2D_coefficients(hdr, fit_inv_poly_u, 'AP', keeplinear=True)
+                _store_2D_coefficients(hdr, fit_inv_poly_v, 'BP', keeplinear=True)
+            hdr['sipmxerr'] = (max_resid * plate_scale, 'Max diff from GWCS (equiv pix).')
+            if max_inv_pix_error:
+                ipdegree = fit_inv_poly_u.degree
+                hdr['ap_order'] = ipdegree
+                hdr['bp_order'] = ipdegree
+                hdr['sipiverr'] = (max_inv_resid, 'Max diff for inverse (pixels)')
+        else:
+            hdr['ctype1'] = 'RA---TAN'
+            hdr['ctype2'] = 'DEC--TAN'
+
+        hdr['cd1_1'] = cdmat[0][0]
+        hdr['cd1_2'] = cdmat[0][1]
+        hdr['cd2_1'] = cdmat[1][0]
+        hdr['cd2_2'] = cdmat[1][1]
+        return hdr
+
+
+def _fit_2D_poly(ntransform, npoints, degree, max_error,
+                 xin, yin, xout, yout,
+                 xind, yind, xoutd, youtd,
+                 verbose=False):
+    """
+    Fit a pair of ordinary 2D polynomial to the supplied transform.
+
+    Use adaptive sampling based on the current degree.
+    """
+    llsqfitter = LinearLSQFitter()
+
+    # The case of one pass with the specified polynomial degree
+    if degree:
+        deglist = [degree]
+    else:
+        deglist = range(10)
+    prev_max_error = -1
+    if verbose:
+        print(f'maximum_specified_error: {max_error}')
+    for deg in deglist:
+        poly_x = Polynomial2D(degree=deg)
+        poly_y = Polynomial2D(degree=deg)
+        fit_poly_x = llsqfitter(poly_x, xin, yin, xout)
+        fit_poly_y = llsqfitter(poly_y, xin, yin, yout)
+        max_resid = _compute_distance_residual(xout, yout,
+                                               fit_poly_x(xin, yin),
+                                               fit_poly_y(xin, yin))
+        if prev_max_error < 0:
+            prev_max_error = max_resid
+        else:
+            if max_resid > prev_max_error:
+                raise RuntimeError('Failed to achieve required error tolerance')
+        if verbose:
+            print(f'Degree = {deg}, max_resid = {max_resid}')
+        if max_resid < max_error:
+            # Check to see if double sampling meets error requirement.
+            max_resid = _compute_distance_residual(xoutd, youtd,
+                                                   fit_poly_x(xind, yind),
+                                                   fit_poly_y(xind, yind))
+            if verbose:
+                print(f'Double sampling check: maximum residual={max_resid}')
+            if max_resid < max_error:
+                if verbose:
+                    print('terminating condition met')
+                break
```

Comment:
I think this code could be rewritten using `while-else`:
```suggestion
        while max_resid >= max_error:
            # Check to see if double sampling meets error requirement.
            max_resid = _compute_distance_residual(xoutd, youtd,
                                                   fit_poly_x(xind, yind),
                                                   fit_poly_y(xind, yind))
            if verbose:
                print(f'Double sampling check: maximum residual={max_resid}')
        else:
            if verbose:
                print('terminating condition met')
```

Reply:
Looks to me that the break is still needed. I don't know if this really makes it much simpler and is a little misleading when a while  is used (one usually thinks that it will iterate until finished when in this case it is designed only to run once). Finally, doesn't it iterate forever if the double sampling leads to large an error?

Reply:
Yeah, disregard this. I do not remember what I was thinking, but probably I was believing you *will be* using a truly adaptive grid size. I probably forgot to remove this once I realized that doubling of the grid size has no real use actually (IMO).

Reply:
From what I've seen, the max error is usually at the boundary and in that case doubling makes no difference. In poorly sampled cases, that may not be true.